src/Compilers/Core/Portable/Operations/Operation.cs

```diff
@@ -46,41 +76,203 @@ public Operation(OperationKind kind, SyntaxNode syntax, ITypeSymbol type, Option
 
         public abstract TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument);
 
-        public static IOperation CreateOperationNone(SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
+        protected void SetParentOperation(IOperation parent)
         {
-            return new NoneOperation(node, constantValue, getChildren);
+            var result = Interlocked.CompareExchange(ref _parentDoNotAccessDirectly, parent, null);
+
+            // tree must belong to same semantic model if parent is given
+            Debug.Assert(parent == null || ((Operation)parent).SemanticModel == SemanticModel);
+
+            // make sure given parent and one we already have is same if we have one already
+            Debug.Assert(result == null || result == parent);
         }
 
-        private class NoneOperation : IOperation
+        /// <summary>
+        /// Create <see cref="IOperation"/> of <see cref="OperationKind.None"/> with explicit children
+        /// 
+        /// Use this to create IOperation when we don't have proper specific IOperation yet for given language construct
+        /// </summary>
+        public static IOperation CreateOperationNone(SemanticModel semanticModel, SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
         {
-            private readonly Func<ImmutableArray<IOperation>> _getChildren;
+            return new NoneOperation(semanticModel, node, constantValue, getChildren);
+        }
+
+        public static T SetParentOperation<T>(T operation, IOperation parent) where T : IOperation
+        {
+            // explicit cast is not allowed, so using "as" instead
+            (operation as Operation)?.SetParentOperation(parent);
+            return operation;
+        }
 
-            public NoneOperation(SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
+        public static ImmutableArray<T> SetParentOperation<T>(ImmutableArray<T> operations, IOperation parent) where T : IOperation
+        {
+            // check quick bail out case first
+            if (operations.Length == 0)
             {
-                Syntax = node;
-                ConstantValue = constantValue;
-                _getChildren = getChildren;
+                // no element
+                return operations;
+            }
+
+            // race is okay. penalty is going through a loop one more time
+            // explicit cast is not allowed, so using "as" instead
+            // invalid expression can have null element in the array
+            if ((operations[0] as Operation)?._parentDoNotAccessDirectly != null)
+            {
+                // already initialized
```

Comment:
@cston asked same thing :) here are the link - https://github.com/dotnet/roslyn/pull/20964#pullrequestreview-52909830

but sure, I will update the comment.